base/src/com/google/idea/blaze/base/model/ExternalWorkspaceQuerySync.java

```diff
@@ -0,0 +1,108 @@
+import com.google.idea.blaze.base.settings.BlazeImportSettings;
+import com.google.idea.blaze.base.sync.SyncScope;
+import com.google.idea.blaze.base.sync.SyncScope.SyncCanceledException;
+import com.google.idea.blaze.base.sync.SyncScope.SyncFailedException;
+import com.google.idea.blaze.exception.BuildException;
+import com.intellij.openapi.project.Project;
+
+import java.util.List;
+
+public class ExternalWorkspaceQuerySync {
+
+  public static void syncExternalWorkspace(Project project, BlazeContext context, ProjectViewSet projectViewSet, BlazeProjectData blazeProjectData) {
+    if (Blaze.getProjectType(project) != BlazeImportSettings.ProjectType.QUERY_SYNC) {
+      return;
+    }
+
+    BuildSystem.BuildInvoker invoker =
+        Blaze.getBuildSystemProvider(project).getBuildSystem().getBuildInvoker(project, context);
+    try {
+      var data = getExternalWorkspaceData(
+          project,
+          context,
+          projectViewSet,
+          blazeProjectData.getBlazeVersionData(),
+          invoker.getBlazeInfo()
+      );
+      ExternalWorkspaceDataManager.getInstance(project).setData(data);
+    } catch (SyncCanceledException | SyncFailedException e) {
+      throw new RuntimeException(e);
+    }
+
+  }
+
+  private static ExternalWorkspaceData getExternalWorkspaceData(
+      Project project,
+      BlazeContext context,
+      ProjectViewSet projectViewSet,
+      BlazeVersionData blazeVersionData,
+      BlazeInfo blazeInfo)
+      throws SyncScope.SyncCanceledException, SyncScope.SyncFailedException {
+
+    List<String> blazeModFlags =
+        BlazeFlags.blazeFlags(
+            project,
+            projectViewSet,
+            BlazeCommandName.MOD,
+            context,
+            BlazeInvocationContext.SYNC_CONTEXT);
+
+    ListenableFuture<ExternalWorkspaceData> externalWorkspaceDataFuture =
+        ExternalWorkspaceDataProvider.getInstance(project)
+            .getExternalWorkspaceData(context, blazeModFlags, blazeVersionData, blazeInfo);
+
+    FutureResult<ExternalWorkspaceData> externalWorkspaceDataResult =
+        FutureUtil.waitForFuture(context, externalWorkspaceDataFuture)
+            .timed(Blaze.buildSystemName(project) + "Mod", EventType.BlazeInvocation)
+            .withProgressMessage("Resolving module repository mapping...")
+            .onError(String.format("Could not run %s mod dump_repo_mapping", Blaze.buildSystemName(project)))
+            .run();
+
+    ExternalWorkspaceData externalWorkspaceData = externalWorkspaceDataResult.result();
+    if (externalWorkspaceData == null) {
+      Exception exception = externalWorkspaceDataResult.exception();
+      if (exception != null) {
+        Throwable cause = exception.getCause();
+        if (cause instanceof BuildException
+            && cause.getCause() instanceof ExecutionDeniedException) {
+          throw new SyncCanceledException();
+        }
+      }
+      throw new SyncFailedException();
```

Comment:
Also if we don't deal with theses exceptions later on anyway, there is no need to differentiate theses cases. Just propagate a `BuildException` here too.